### PR TITLE
QS-229 add etl status to 'plextrac check' command

### PR DESCRIPTION
**Pull request by rkeeler**
Testing steps to come soonTM

## Example Output
**Note: the raw json only prints if you pass `-v` as a cli flag**

```
plextrac check -v

<output from other checks omitted>

-- Checking CB->PG ETL status --------------------------------------------------

!!! One or more ETLs are in an unhealthy status.

    Domains (Read Mode / Write Mode):
      authentication: couchbase_only / couchbase_only
                 ctf: couchbase_only / couchbase_only
              tenant: couchbase_only / couchbase_only
      tenantSettings: couchbase_only / couchbase_only
                user: couchbase_only / couchbase_only
    ETLs:
      etl_tenant_v1: UNKNOWN
      etl_user_v1: SUCCESS

    {
      "domains": {
        "authentication": {
          "featureGateName": "PG_MIGRATE_USER",
          "featureGateStatus": "disabled",
          "etlStatus": {},
          "readMode": "couchbase_only",
          "writeMode": "couchbase_only"
        },
        "ctf": {
          "featureGateName": "PG_MIGRATE_USER",
          "featureGateStatus": "disabled",
          "etlStatus": {},
          "readMode": "couchbase_only",
          "writeMode": "couchbase_only"
        },
        "tenant": {
          "featureGateName": "PG_MIGRATE_TENANT",
          "featureGateStatus": "enabled",
          "etlStatus": {
            "etl_tenant_v1": "UNKNOWN"
          },
          "readMode": "couchbase_only",
          "writeMode": "couchbase_only"
        },
        "tenantSettings": {
          "featureGateName": "PG_MIGRATE_TENANT",
          "featureGateStatus": "enabled",
          "etlStatus": {
            "etl_tenant_v1": "UNKNOWN"
          },
          "readMode": "couchbase_only",
          "writeMode": "couchbase_only"
        },
        "user": {
          "featureGateName": "PG_MIGRATE_USER",
          "featureGateStatus": "disabled",
          "etlStatus": {},
          "readMode": "couchbase_only",
          "writeMode": "couchbase_only"
        }
      },
      "etls": {
        "etl_tenant_v1": {
          "status": "UNKNOWN",
          "error": "Could not read etl result file."
        },
        "etl_user_v1": {
          "status": "SUCCESS",
          "error": "Could not read etl result file."
        }
      },
      "etlsCombinedStatus": "HEALTHaY"
    }
```

### CODE REVIEW

File: src/_check.sh [hunk 1]
```diff
@@ -128,3 +128,38 @@ function _check_base_required_packages() {
   fi
   return $status
 }
+
+function check_etl_status() {
+  title "Checking CB->PG ETL status"
+
+  RAW_OUTPUT=$(`compose_client exec plextracapi npm run pg:etl:status`)
```
Comment (rkeeler) [inline src/_check.sh:135]: This npm script has not been merged to main yet. I'm going to leave this PR in draft until I get that script merged into the backend repo first.

**Review by jesseops — COMMENTED** [review]
Overall looks good 👍🏼 but might have an easier way to sanitize the output from the NPM script

File: src/_check.sh [hunk 2]
```diff
@@ -128,3 +128,38 @@ function _check_base_required_packages() {
...
+"
+  SUMMARY_MARKER=""$NEW_LINE"---------------- SUMMARY ----------------"$NEW_LINE""
+  JSON_MARKER=""$NEW_LINE"---------------- RAW JSON ----------------"$NEW_LINE""
+  
+  SUMMARY_OUTPUT="${RAW_OUTPUT#*"$SUMMARY_MARKER"}" # trim everything before SUMMARY marker 
+
+  JSON_OUTPUT="${RAW_OUTPUT#*"$JSON_MARKER"}" # trim everything before RAW JSON marker
+  JSON_OUTPUT="${JSON_OUTPUT%"$SUMMARY_MARKER"*}" # trim everything after SUMMARY marker
+  
+  ETLS_COMBINED_STATUS=$(echo $JSON_OUTPUT | jq -r .etlsCombinedStatus)
```
[the remark this answers is not in the document]
Comment (jesseops) [reply src/_check.sh:154]: (I yoinked this stuff from stackoverflow 😂 )

[the remark this answers is not in the document]
Comment (rkeeler) [reply src/_check.sh:147]: Sweet, thanks for the pointers. I was able to tweak what you suggested a bit and find a couple of sed commands that select the json content and everything after the json content separately. No more need for the marker lines in the text even.

File: src/_check.sh [hunk 3]
```diff
@@ -120,3 +121,29 @@ function _check_base_required_packages() {
...
+
+  if [[ "$ETLS_COMBINED_STATUS" == "HEALTHY" ]] 
+  then
+    info "All ETLs are in a healthy status."
+  else
+    error "One or more ETLs are in an unhealthy status."
+  fi
+
+  msg "$SUMMARY_OUTPUT\n"
```
Comment (jesseops) [inline src/_check.sh:147]: I remembered this looking over a different PR, msg should be log for correct formatting.
```suggestion
  log "$SUMMARY_OUTPUT\n"
```

File: src/_check.sh [hunk 4]
```diff
@@ -120,3 +121,29 @@ function _check_base_required_packages() {
...
+  debug "$JSON_OUTPUT\n"
+}
```
Comment (jesseops) [inline src/_check.sh:149]:
```suggestion
  title "Checking Couchbase-to-Postgres ETL status"

  etlStatusOut=$(compose_client exec plextracapi npm run pg:etl:status 2>&1) || true # don't break if the script exits unhealthy

  # Find the json content by looking for the first line that starts
  # with an opening brace and the first line that starts with a closing brace.
  etlStatusJson=$(echo "$etlStatusOut" | sed '/^{/,/^}/!d')

  # Find the summary content by finding the first line that starts
  # with a closing brace and selecting all remaining lines after that one.
  etlStatusSummary=$(echo "$etlStatusOut" | sed '1,/^}/d')

  etlStatusCombinedState=$(echo $etlStatusJson | jq -r .etlsCombinedStatus)

  if [[ "$etlStatusCombinedState" == "HEALTHY" ]] 
  then
    info "All ETLs are in a healthy status."
  else
    error "One or more ETLs are in an unhealthy status."
  fi

  msg "$etlStatusSummary\n"
  debug "$etlStatusJson\n"
}
```

I have a few "mixed" formats here, but in general internal variables should not be all uppercase in order to differentiate from environment variables and avoid potential conflicts. https://unix.stackexchange.com/a/42849

**Pull request closed without merging**